rotkehlchen/chain/solana/modules/jupiter/decoder.py

```diff
@@ -30,118 +27,153 @@
-        out_token_amounts: defaultdict[Asset, set[FVal]] = defaultdict(set)
+        # Find the route instruction corresponding to this swap event instruction
         for instruction in context.transaction.instructions:
-            if (
-                    instruction.parent_execution_index != context.instruction.execution_index or
-                    instruction.data[:SWAP_EVENT_DISCRIMINATOR_LEN] != SWAP_EVENT_DISCRIMINATOR
+            if instruction.execution_index == context.instruction.parent_execution_index and (
+                (is_v2 := match_discriminator(instruction.data, ROUTE_V2_DISCRIMINATOR)) or
+                match_discriminator(instruction.data, ROUTE_DISCRIMINATOR)
             ):
-                continue
-
-            try:
-                decoded_event = SWAP_EVENT_LAYOUT.parse(instruction.data[SWAP_EVENT_DISCRIMINATOR_LEN:])  # skip 16-byte discriminator  # noqa: E501
-            except ConstructError as e:
-                log.error(
-                    f'Failed to parse Jupiter swap event data in transaction '
-                    f'{context.transaction.signature} at execution_index {instruction.execution_index} '  # noqa: E501
-                    f'(parent: {instruction.parent_execution_index}): {e}',
-                )
-                continue
+                destination_mint = instruction.accounts[4] if is_v2 else instruction.accounts[5]
+                route_instruction = instruction
+                break
+        else:
+            log.error(f'Failed to find Jupiter route instruction in transaction {context.transaction!s}')  # noqa: E501
+            return DEFAULT_SOLANA_DECODING_OUTPUT
 
-            for token_amounts, mint_address_bytes, raw_amount in [
-                (out_token_amounts, decoded_event.inputMint, decoded_event.inputAmount),
-                (in_token_amounts, decoded_event.outputMint, decoded_event.outputAmount),
-            ]:
-                token_amounts[token := get_or_create_solana_token(
-                    userdb=self.node_inquirer.database,
-                    address=bytes_to_solana_address(mint_address_bytes),
-                    solana_inquirer=self.node_inquirer,
-                )].add(token_normalized_value(token=token, token_amount=raw_amount))
-
-        out_event, in_event, filtered_events = None, None, []
+        unrelated_events, other_events, platform_fee_event = [], [], None
+        out_events_by_asset: dict[Asset, SolanaEvent] = {}
+        in_events_by_asset: dict[Asset, SolanaEvent] = {}
         for event in context.decoded_events:
             if (
-                event.event_type == HistoryEventType.SPEND and
-                event.event_subtype == HistoryEventSubType.NONE and
-                event.amount in out_token_amounts[event.asset]
+                event.instruction is None or
+                event.instruction.parent_execution_index != route_instruction.execution_index
```

Comment:
instead of storing the instruction in `SolanaEvent`, we could add a method to the base decoder tools class that generates `SolanaEvent.sequence_index` from the `parent_execution_index` and `execution_index` combination. then add a separate method to check whether an instruction generated the event.